rasa_core/training/structures.py

```diff
@@ -274,24 +284,89 @@ def with_cycles_removed(self):
                     for cp in step.start_checkpoints:
                         if cp.name == original_cp:
                             if k == e:
-                                cid = source_cid
+                                cp_name = source_cp_name
                             else:
-                                cid = connector_cid
+                                cp_name = connector_cp_name
                                 needs_connector = True
 
-                            additional_ends.append(Checkpoint(cid,
-                                                              cp.conditions))
+                            if not self._is_checkpoint_in_list(
+                                    cp_name, cp.conditions,
+                                    step.start_checkpoints):
+                                # add checkpoint only if it was not added
+                                additional_ends.append(
+                                        Checkpoint(cp_name, cp.conditions))
+
                 if additional_ends:
                     updated = step.create_copy(use_new_id=False)
                     updated.start_checkpoints.extend(additional_ends)
                     story_steps[k] = updated
 
             if needs_connector:
-                start.end_checkpoints.append(Checkpoint(connector_cid))
+                start.end_checkpoints.append(Checkpoint(connector_cp_name))
+
+        self._remove_unused_generated_cps(story_steps,
+                                          all_overlapping_cps,
+                                          story_end_checkpoints)
 
-        return StoryGraph(story_steps.values(),
+        return StoryGraph(list(story_steps.values()),
                           story_end_checkpoints)
 
+    @staticmethod
+    def _checkpoint_difference(cps, cp_to_ignore):
+        return [cp for cp in cps if cp.name not in cp_to_ignore]
+
+    def _remove_unused_generated_cps(self, story_steps, overlapping_cps,
+                                     story_end_checkpoints):
+        # the process above may generate unused start checkpoints
+        # we need to find them and remove them
+        # also there might be generated unused end checkpoints
+        unused_cps = self._unused_checkpoints(story_steps.values(),
+                                              story_end_checkpoints)
+
+        unused_overlapping_cps = unused_cps.intersection(overlapping_cps)
+
+        unused_genr_cps = {cp_name
+                           for cp_name in unused_cps
+                           if cp_name.startswith(GENERATED_CHECKPOINT_PREFIX)}
+
+        for k, step in story_steps.items():
+            # changed all ends
+            updated = step.create_copy(use_new_id=False)
```

Comment:
why do we need to copy this if we replace it anways?

Reply:
do not know, i just used the previous code here.

Reply:
ok :+1:

Reply:
I guess because this step might be linked to another place in the list?

Reply:
let's just keep it this way.